File: WinPaletter/GUI/Elements/Trackbars/ColorBar.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text

Namespace UI.WP

    <Description("Themed Trackbar for selecting colors (HSL) for WinPaletter UI")>
    <DefaultEvent("Scroll")> Public Class ColorBar : Inherits Control

        Event Scroll(sender As Object)

#Region "Properties"
        Private _Minimum As Integer
        Property Minimum() As Integer
            Get
                Return _Minimum
            End Get
            Set(value As Integer)
                _Minimum = value
                If value > _Value Then _Value = value
                If value > _Maximum Then _Maximum = value

                InvalidateLayout()
            End Set
        End Property

        Private _Maximum As Integer = 100
        Property Maximum() As Integer
            Get
                Return _Maximum
            End Get
            Set(value As Integer)
                _Maximum = value
                If value < _Value Then _Value = value
                If value < _Minimum Then _Minimum = value
                InvalidateLayout()
            End Set
        End Property

        Private _Value As Integer
        Property Value() As Integer
            Get

                Return _Value
            End Get
            Set(value As Integer)
                If value = _Value Then Return

                If value > _Maximum Then
                    value = _Maximum
                End If

                If value < _Minimum Then
                    value = _Minimum
                End If

                _Value = value
                InvalidatePosition()

                RaiseEvent Scroll(Me)
            End Set
        End Property

        Private _SmallChange As Integer = 1
        Public Property SmallChange() As Integer
            Get
                Return _SmallChange
            End Get
            Set(value As Integer)
                If value < 1 Then
                    Throw New Exception("Property value is not valid.")
                End If

                _SmallChange = value
            End Set
        End Property

        Private _LargeChange As Integer = 10
        Public Property LargeChange() As Integer
            Get
                Return _LargeChange
            End Get
            Set(value As Integer)
                If value < 1 Then
                    Throw New Exception("Property value is not valid.")
                End If

                _LargeChange = value
            End Set
        End Property

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String = ""
#End Region

        Enum ModesList
            Hue
            Saturation
            Light
        End Enum

        Public Property Mode As ModesList = ModesList.Hue
        ReadOnly ButtonSize As Integer = 0
        ReadOnly ThumbSize As Integer = 35 ' 14 minimum
        Private LSA As Rectangle
        Private RSA As Rectangle
        Private Shaft As Rectangle
        Private Thumb As Rectangle
        Private ThumbDown As Boolean
        Private Circle As Rectangle
        Private _Shown As Boolean = False

        Enum MouseState
            None
            Over
            Down
        End Enum

#Region "Animator"
        Dim alpha As Integer
        ReadOnly Factor As Integer = 25
        Dim WithEvents Tmr As New Timer With {.Enabled = False, .Interval = 1}

        Private Sub Tmr_Tick(sender As Object, e As EventArgs) Handles Tmr.Tick
            If Not DesignMode Then

                If State = MouseState.Over And _Shown Then
                    If alpha + Factor <= 255 Then
                        alpha += Factor
                    ElseIf alpha + Factor > 255 Then
                        alpha = 255
                        Tmr.Enabled = False
                        Tmr.Stop()
                    End If

                    Threading.Thread.Sleep(1)
                    Invalidate()
                End If

                If _Shown And (Not State = MouseState.Over Or State = MouseState.Down) Then
                    If alpha - Factor >= 0 Then
                        alpha -= Factor
                    ElseIf alpha - Factor < 0 Then
                        alpha = 0
                        Tmr.Enabled = False
                        Tmr.Stop()
                    End If

                    Threading.Thread.Sleep(1)
                    Invalidate()
                End If
            End If
        End Sub
#End Region

        Public State As MouseState = MouseState.None

        Sub New()
            SetStyle(DirectCast(139286, ControlStyles), True)
            SetStyle(ControlStyles.Selectable, False)
            Height = 19
            Text = ""
        End Sub

        Dim I1 As Integer

        Private _AccentColor As Color = My.Style.Colors.Core
        Public Property AccentColor As Color
            Get
                Return _AccentColor
            End Get
            Set(value As Color)
                _AccentColor = value
                Refresh()
            End Set
        End Property


        Private _H As Integer = 0
        Public Property H As Integer
            Get
                Return _H
            End Get
            Set(value As Integer)
                _H = value
                Refresh()
            End Set
        End Property


        Private _S As Single = 1
        Public Property S As Single
            Get
                Return _S
            End Get
            Set(value As Single)
                _S = value
                Refresh()
            End Set
        End Property


        Private _L As Integer = 0.5
        Public Property L As Single
            Get
                Return _L
            End Get
            Set(value As Single)
                _L = value
                Refresh()
            End Set
        End Property

        Private ReadOnly Color1 As Color = Color.FromArgb(255, 23, 0)
        Private ReadOnly Color2 As Color = Color.FromArgb(253, 253, 0)
        Private ReadOnly Color3 As Color = Color.FromArgb(58, 255, 0)
        Private ReadOnly Color4 As Color = Color.FromArgb(0, 255, 240)
        Private ReadOnly Color5 As Color = Color.FromArgb(0, 17, 255)
        Private ReadOnly Color6 As Color = Color.FromArgb(255, 0, 212)
        Private ReadOnly Color7 As Color = Color.FromArgb(255, 0, 23)

        ReadOnly cb_H As New ColorBlend() With {.Positions = {0, 1 / 6.0F, 2 / 6.0F, 3 / 6.0F, 4 / 6.0F, 5 / 6.0F, 1},
                        .Colors = {Color1, Color2, Color3, Color4, Color5, Color6, Color7}}

        Dim cb_L As New ColorBlend() With {.Positions = {0, 1 / 2.0F, 1}, .Colors = {Color.Black, _AccentColor, Color.White}}

        Protected Overrides Sub OnPaint(e As System.Windows.Forms.PaintEventArgs)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.HighQuality
            G.TextRenderingHint = TextRenderingHint.AntiAliasGridFit

            G.Clear(BackColor)
            Dim Dark As Boolean = My.Style.DarkMode
            Dim c_back As Color = If(Dark, Color.FromArgb(60, 60, 60), Color.FromArgb(210, 210, 210))
            Dim c_btn As Color = If(Dark, Color.FromArgb(165, 165, 165), Color.FromArgb(100, 100, 100))
            Dim C As Color

            Dim middleRect As New Rectangle(0, (Height - (Height * 0.25)) / 2, Width - 1, Height * 0.25)

            Dim back As LinearGradientBrush

            Select Case Mode
                Case ModesList.Hue
                    back = New LinearGradientBrush(middleRect, Color.Black, Color.Black, 0, False) With {.InterpolationColors = cb_H}
                    Dim HSL_ As HSL_Structure = Color.FromArgb(0, 255, 240).ToHSL()
                    HSL_.H = (Value / Maximum) * 359
                    C = HSL_.ToRGB

                Case ModesList.Saturation
                    Dim HSL_x1 As HSL_Structure = _AccentColor.ToHSL()
                    Dim HSL_x2 As HSL_Structure = _AccentColor.ToHSL()
                    HSL_x1.S = 0
                    HSL_x2.S = 1
                    back = New LinearGradientBrush(middleRect, HSL_x1.ToRGB, HSL_x2.ToRGB, LinearGradientMode.Horizontal)
                    Dim HSL_ As HSL_Structure = _AccentColor.ToHSL()
                    HSL_.S = Value / Maximum
                    C = HSL_.ToRGB

                Case ModesList.Light
                    cb_L = New ColorBlend() With {.Positions = {0, 1 / 2.0F, 1}, .Colors = {Color.Black, _AccentColor, Color.White}}
                    back = New LinearGradientBrush(middleRect, Color.Black, Color.Black, 0, False) With {.InterpolationColors = cb_L}
                    Dim HSL_ As HSL_Structure = _AccentColor.ToHSL()
                    HSL_.L = Value / Maximum
                    C = HSL_.ToRGB

                Case Else
                    C = If(Dark, Color.FromArgb(60, 60, 60), Color.FromArgb(210, 210, 210))
                    back = New LinearGradientBrush(middleRect, C, C, GradientMode.Horizontal)

            End Select

            G.FillRoundedRect(back, middleRect)

            Circle = New Rectangle((Value / Maximum) * Shaft.Width, 0, Height - 1, Height - 1)

            Using br As New SolidBrush(BackColor) : G.FillRectangle(br, New Rectangle(-1, 0, 4, Height)) : End Using

            Using br As New SolidBrush(BackColor) : G.FillRectangle(br, New Rectangle(Width - 4, 0, 4, Height)) : End Using

            Using br As New SolidBrush(My.Style.Colors.Border) : G.FillEllipse(br, Circle) : End Using

            Dim smallC1 As New Rectangle(Circle.X + 5, Circle.Y + 5, Circle.Width - 10, Circle.Height - 10)
            Dim smallC2 As New Rectangle(Circle.X + 4, Circle.Y + 4, Circle.Width - 8, Circle.Height - 8)

            Using br As New SolidBrush(C) : G.FillEllipse(br, smallC1) : End Using
            Using br As New SolidBrush(Color.FromArgb(alpha, C)) : G.FillEllipse(br, smallC2) : End Using
        End Sub

        Protected Overrides Sub OnSizeChanged(e As EventArgs)
            Height = 19
            InvalidateLayout()
        End Sub

        Private Sub InvalidateLayout()
            LSA = New Rectangle(0, 0, ButtonSize, Height)
            RSA = New Rectangle(Width - ButtonSize, 0, ButtonSize, Height)
            Shaft = New Rectangle(LSA.Right + 1 + 0.5 * Height, 0, Width - Height - 1, Height)
            Thumb = New Rectangle(0, 1, (Value / Maximum) * Shaft.Width, Height - 3)
            Circle = New Rectangle((Value / Maximum) * Shaft.Width, 0, Height - 1, Height - 1)
            RaiseEvent Scroll(Me)
            InvalidatePosition()
        End Sub

        Private Sub InvalidatePosition()
            Thumb.Width = (Value / Maximum) * Width
            Refresh()
        End Sub

        Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
            If e.Button = Windows.Forms.MouseButtons.Left Then
                State = MouseState.Down
                Tmr.Enabled = True
                Tmr.Start()

                If Circle.Contains(e.Location) Then
                    ThumbDown = True
                    Return
                Else
                    If e.X < Circle.X Then

                        I1 = _Value - _LargeChange
                    Else
                        I1 = _Value + _LargeChange
                    End If
                End If

                Value = Math.Min(Math.Max(I1, _Minimum), _Maximum)

                InvalidatePosition()
            End If
        End Sub

        Protected Overrides Sub OnMouseMove(e As MouseEventArgs)
            If Circle.Contains(e.Location) And Not e.Button = MouseButtons.Left Then
                State = MouseState.Over
            Else
                If e.Button = MouseButtons.Left Then State = MouseState.Down Else State = MouseState.None
            End If

            Invalidate()

            If ThumbDown Then
                Value = Math.Min(Math.Max((e.X / Width) * Maximum, _Minimum), _Maximum)
                InvalidatePosition()
            End If

            Tmr.Enabled = True
            Tmr.Start()
        End Sub

        Private Sub NSVScrollBar_MouseUp(sender As Object, e As MouseEventArgs) Handles Me.MouseUp
            ThumbDown = False
            State = MouseState.None
            Tmr.Enabled = True
            Tmr.Start()
            Invalidate()
        End Sub

        Private Sub XenonScrollBarV_MouseEnter(sender As Object, e As EventArgs) Handles Me.MouseEnter
            If Thumb.Contains(MousePosition) Then
                State = MouseState.Over
                Invalidate()
                Tmr.Enabled = True
                Tmr.Start()
            End If
        End Sub

        Private Sub XenonScrollBarV_MouseLeave(sender As Object, e As EventArgs) Handles Me.MouseLeave
            State = MouseState.None
            Tmr.Enabled = True
            Tmr.Start()
            Invalidate()
        End Sub

        Private Sub XenonScrollBarV_MouseWheel(sender As Object, e As MouseEventArgs) Handles Me.MouseWheel
            If e.Delta < 0 Then
                If Value < Maximum Then
                    If e.Delta <= -240 Then Value += LargeChange Else Value += SmallChange
                End If
            Else
                If Value > Minimum Then
                    If e.Delta >= 240 Then Value -= LargeChange Else Value -= SmallChange
                End If
            End If
        End Sub

        Private Sub XenonRadioButton_HandleCreated(sender As Object, e As EventArgs) Handles Me.HandleCreated

            Try
                If Not DesignMode Then
                    AddHandler FindForm.Load, AddressOf Loaded
                    AddHandler FindForm.Shown, AddressOf Showed
                    AddHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                    AddHandler Parent.VisibleChanged, AddressOf RefreshColorPalette
                    AddHandler Parent.EnabledChanged, AddressOf RefreshColorPalette
                    AddHandler VisibleChanged, AddressOf RefreshColorPalette
                    AddHandler EnabledChanged, AddressOf RefreshColorPalette
                End If
            Catch
            End Try

            Try
                alpha = 0

            Catch
            End Try
        End Sub

        Private Sub XenonColorBar_HandleDestroyed(sender As Object, e As EventArgs) Handles Me.HandleDestroyed
            Try
                If Not DesignMode Then
                    RemoveHandler FindForm.Load, AddressOf Loaded
                    RemoveHandler FindForm.Shown, AddressOf Showed
                    RemoveHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                    RemoveHandler Parent.VisibleChanged, AddressOf RefreshColorPalette
                    RemoveHandler Parent.EnabledChanged, AddressOf RefreshColorPalette
                    RemoveHandler VisibleChanged, AddressOf RefreshColorPalette
                    RemoveHandler EnabledChanged, AddressOf RefreshColorPalette
                End If
            Catch
            End Try
        End Sub

        Sub Loaded()
            _Shown = False
        End Sub

        Sub Showed()
            _Shown = True

            Invalidate()
        End Sub

        Public Sub RefreshColorPalette()

            Invalidate()
        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Trackbars/ColorBar.cs
using FluentTransitions;
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WinPaletter.UI.WP
{
    [Description("Themed Trackbar for selecting colors (HSL) for WinPaletter UI")]
    [DefaultEvent("Scroll")]
    public class ColorBar : Control
    {
        public ColorBar()
        {
            cb_H = new()
            {
                Positions = [0f, 1f / 6.0f, 2f / 6.0f, 3f / 6.0f, 4f / 6.0f, 5f / 6.0f, 1f],
                Colors = [Color1, Color2, Color3, Color4, Color5, Color6, Color7]
            };
            cb_L = new() { Positions = [0f, 0.5f, 1f], Colors = [Color.Black, _AccentColor, Color.White] };

            SetStyle(ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor | (ControlStyles)139286, true);
            SetStyle(ControlStyles.Selectable, true);
            DoubleBuffered = true;
            BackColor = Color.Transparent;

            Text = string.Empty;
            _alpha = 0;

            TabStop = true; // allow keyboard focus
        }

        #region Variables
        private bool CanAnimate => !DesignMode && Program.Style.Animations && this != null && Visible && Parent?.Visible == true && FindForm()?.Visible == true;

        private readonly float ButtonSize = 0;
        private RectangleF LSA;
        private RectangleF RSA;
        private RectangleF Shaft;
        private RectangleF Thumb;
        private bool ThumbDown;
        private RectangleF Circle;
        private float I1;

        private readonly Color Color1 = Color.FromArgb(255, 23, 0);
        private readonly Color Color2 = Color.FromArgb(253, 253, 0);
        private readonly Color Color3 = Color.FromArgb(58, 255, 0);
        private readonly Color Color4 = Color.FromArgb(0, 255, 240);
        private readonly Color Color5 = Color.FromArgb(0, 17, 255);
        private readonly Color Color6 = Color.FromArgb(255, 0, 212);
        private readonly Color Color7 = Color.FromArgb(255, 0, 23);
        private readonly ColorBlend cb_H;
        private ColorBlend cb_L;

        public MouseState State = MouseState.None;

        private LinearGradientBrush? backBrushCache;

        public enum MouseState { None, Over, Down }
        public enum ModesList { Hue, Saturation, Light }
        #endregion

        #region Properties
        private Color _AccentColor = Program.Style.Schemes.Main.Colors.AccentAlt;
        public Color AccentColor
        {
            get => _AccentColor;
            set
            {
                if (value != _AccentColor)
                {
                    _AccentColor = value;
                    InvalidateBrushCache();
                    Invalidate();
                }
            }
        }

        private float _H = 0f;
        public float H { get => _H; set { if (value != _H) { _H = value; Invalidate(); } } }

        private float _S = 1f;
        public float S { get => _S; set { if (value != _S) { _S = value; Invalidate(); } } }

        private float _L = 0.5f;
        public float L { get => _L; set { if (value != _L) { _L = value; Invalidate(); } } }

        private int _Minimum;
        public int Minimum
        {
            get => _Minimum;
            set
            {
                if (value != _Minimum)
                {
                    _Minimum = value;
                    if (value > _Maximum) _Maximum = value;
                    if (value > _Value) _Value = value;
                    InvalidateLayout();
                }
            }
        }

        private int _Maximum = 100;
        public int Maximum
        {
            get => _Maximum;
            set
            {
                if (value != _Maximum)
                {
                    _Maximum = value;
                    if (value < _Value) _Value = value;
                    if (value < _Minimum) _Minimum = value;
                    InvalidateLayout();
                }
            }
        }

        private int _Value;
        public int Value
        {
            get => _Value;
            set
            {
                if (value > _Maximum) value = _Maximum;
                if (value < _Minimum) value = _Minimum;

                if (_Value != value)
                {
                    _Value = value;
                    InvalidateLayout();
                    InvalidatePosition();
                    Scroll?.Invoke(this);
                }
            }
        }

        private int _SmallChange = 1;
        public int SmallChange { get => _SmallChange; set => _SmallChange = Math.Max(1, value); }

        private int _LargeChange = 10;
        public int LargeChange { get => _LargeChange; set => _LargeChange = Math.Max(1, value); }

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        public override string Text { get; set; } = string.Empty;

        private ModesList _mode = ModesList.Hue;
        public ModesList Mode
        {
            get => _mode;
            set
            {
                if (value != _mode)
                {
                    _mode = value;
                    InvalidateBrushCache();
                    Invalidate();
                }
            }
        }
        #endregion

        #region Events
        public event ScrollEventHandler? Scroll;
        public delegate void ScrollEventHandler(object sender);
        #endregion

        #region Layout

        private void InvalidateLayout()
        {
            LSA = new(0, 0, ButtonSize, Height);
            RSA = new(Width - ButtonSize, 0, ButtonSize, Height);
            Shaft = new(LSA.Right + 1 + 0.5f * Height, 0, Width - Height - 1, Height);
            Thumb = new(0, 1, Value / Maximum * Shaft.Width, Height - 3);
            Circle = new(Value / Maximum * Shaft.Width, 0, Height - 1, Height - 1);
            Scroll?.Invoke(this);
            InvalidatePosition();
        }

        private async void InvalidatePosition()
        {
            Thumb.Width = Value / Maximum * Width;
            await Task.Delay(10);
            Invalidate();
        }
        #endregion

        #region Mouse / Focus
        protected override void OnMouseDown(MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                State = MouseState.Down;
                if (CanAnimate) Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));
                else alpha = 0;

                if (Circle.Contains(e.Location)) { ThumbDown = true; return; }
                I1 = e.X < Circle.X ? _Value - (float)_LargeChange : _Value + (float)_LargeChange;
                Value = (int)Math.Min(Math.Max(I1, _Minimum), _Maximum);
                InvalidatePosition();
            }

            base.OnMouseDown(e);
        }

        protected override void OnMouseMove(MouseEventArgs e)
        {
            if (Circle.Contains(e.Location) && e.Button != MouseButtons.Left) State = MouseState.Over;
            else if (e.Button == MouseButtons.Left) State = MouseState.Down;
            else State = MouseState.None;

            if (ThumbDown)
            {
                Value = (int)Math.Min(Math.Max(e.X / (float)Width * Maximum, _Minimum), _Maximum);
                InvalidatePosition();
            }

            if (CanAnimate)
                Transition.With(this, nameof(alpha), Circle.Contains(PointToClient(MousePosition)) ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));
            else alpha = Circle.Contains(PointToClient(MousePosition)) ? 255 : 0;

            Invalidate();
            base.OnMouseMove(e);
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            ThumbDown = false;
            State = MouseState.None;
            if (CanAnimate) Transition.With(this, nameof(alpha), ContainsFocus ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));
            else alpha = ContainsFocus ? 255 : 0;
            base.OnMouseUp(e);
        }
        #endregion

        #region Keyboard
        protected override bool IsInputKey(Keys keyData)
        {
            return keyData is Keys.Left or Keys.Right or Keys.Home or Keys.End || base.IsInputKey(keyData);
        }

        protected override void OnKeyDown(KeyEventArgs e)
        {
            base.OnKeyDown(e);
            switch (e.KeyCode)
            {
                case Keys.Left: Value -= SmallChange; break;
                case Keys.Right: Value += SmallChange; break;
                case Keys.Home: Value = Minimum; break;
                case Keys.End: Value = Maximum; break;
            }
        }
        #endregion

        #region Animator
        private int _alpha = 0;
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public int alpha { get => _alpha; set { _alpha = value; Invalidate(); } }
        #endregion

        #region Brush Cache
        private void InvalidateBrushCache()
        {
            backBrushCache?.Dispose();
            backBrushCache = null;
        }

        private LinearGradientBrush GetBackBrush(RectangleF rect)
        {
            InvalidateBrushCache();

            backBrushCache = Mode switch
            {
                ModesList.Hue => new(rect, Color.Black, Color.Black, 0f, false) { InterpolationColors = cb_H },
                ModesList.Saturation => new(rect, _AccentColor.ToHSL().WithS(0f).ToRGB(), _AccentColor.ToHSL().WithS(1f).ToRGB(), LinearGradientMode.Horizontal),
                ModesList.Light => new(rect, Color.Black, Color.Black, 0f, false) { InterpolationColors = cb_L = new() { Positions = [0f, 0.5f, 1f], Colors = [Color.Black, _AccentColor, Color.White] } },
                _ => new(rect, _AccentColor, _AccentColor, LinearGradientMode.Horizontal)
            };
            return backBrushCache;
        }
        #endregion

        #region Paint
        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;
            InvokePaintBackground(this, e);

            Circle = new((Value / (float)Maximum) * Shaft.Width, 0, Height - 1, Height - 1);

            RectangleF middleRect = new(Circle.Width / 4f, (Height - Height * 0.25f) / 2f, Width - (Circle.Width / 4f) * 2f - 1, Height * 0.25f);
            using LinearGradientBrush back = GetBackBrush(middleRect);

            G.FillRoundedRect(back, middleRect);

            using (SolidBrush br = new(Program.Style.Schemes.Main.Colors.Line(parentLevel))) G.FillEllipse(br, Circle);

            RectangleF smallC1 = new(Circle.X + 5, Circle.Y + 5, Circle.Width - 10, Circle.Height - 10);
            RectangleF smallC2 = new(Circle.X + 4, Circle.Y + 4, Circle.Width - 8, Circle.Height - 8);

            Color color = Mode switch
            {
                ModesList.Hue => _AccentColor.ToHSL().WithH(Value).ToRGB(),
                ModesList.Saturation => _AccentColor.ToHSL().WithS(Value / (float)Maximum).ToRGB(),
                ModesList.Light => _AccentColor.ToHSL().WithL(Value / (float)Maximum).ToRGB(),
                _ => _AccentColor
            };

            using (SolidBrush br = new(color)) G.FillEllipse(br, smallC1);
            using (SolidBrush br = new(Color.FromArgb(alpha, color))) G.FillEllipse(br, smallC2);

            base.OnPaint(e);
        }
        #endregion

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);
            parentLevel = this.Level();
        }
    }
}
